lms/views/decorators/h_api.py

```diff
@@ -122,3 +124,124 @@ def wrapper(request, jwt):  # pylint: disable=too-many-branches
         return wrapped(request, jwt)
 
     return wrapper
+
+
+def create_h_group(wrapped):  # noqa: MC0002
+    """
+    Create a group in h if one doesn't exist already.
+
+    Call the h API to create a group for the LTI course, if one doesn't exist
+    already.
+
+    Use this function as a decorator rather than calling it directly.
+    The wrapped view must take ``request`` and ``jwt`` arguments::
+
+      @view_config(...)
+      @create_h_group
+      def my_view(request, jwt):
+          ...
+
+    """
+    # FIXME: This function doesn't do anything with the ``jwt`` argument,
+    # other than pass it through to the wrapped view (or to the next decorator
+    # on the wrapped view). The ``jwt`` argument has to be here because:
+    #
+    # - This decorator is called by the ``@lti_launch`` decorator (because
+    #   ``@lti_launch`` is always placed above this decorator on decorated views)
+    #   and ``@lti_launch`` passes a ``jwt_token`` argument when it calls this
+    #   decorator.
+    #
+    # - The views that this decorator decorates expect a ``jwt`` argument so this
+    #   decorator has to pass it to them (or rather it has to pass it down to the
+    #   next decorator down in the stack, and it eventually gets passed to the
+    #   view).
+    #
+    # This should all be refactored so that views and view decorators aren't
+    # tightly coupled and arguments don't need to be passed through multiple
+    # decorators to the view.
+    def wrapper(request, jwt):  # pylint: disable=too-many-branches
+        # Only create users for application instances that we've enabled the
+        # auto provisioning features for.
+        try:
+            oauth_consumer_key = request.params["oauth_consumer_key"]
+        except KeyError:
+            raise HTTPBadRequest('Required parameter "oauth_consumer_key" missing from LTI params')
+        enabled_consumer_keys = request.registry.settings["auto_provisioning"]
+        if oauth_consumer_key not in enabled_consumer_keys:
+            return wrapped(request, jwt)
```

Comment:
This check for whether the feature is enabled is now duplicated in two places. Needs to move into a feature flags service. For now I should at least move it into a helper function within this file...

Reply:
Yeah, this confused me at first...but I'm sure we'll factor it out in the future. Given that this pertains to `h` specifically, maybe somewhere in the `h_api` module?